service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/flight/RetrieveAzureResourcesRegionStep.java

```diff
@@ -0,0 +1,242 @@
+import bio.terra.workspace.service.resource.controlled.cloud.azure.vm.ControlledAzureVmResource;
+import bio.terra.workspace.service.resource.controlled.model.ControlledResource;
+import bio.terra.workspace.service.resource.model.WsmResource;
+import bio.terra.workspace.service.resource.model.WsmResourceType;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import bio.terra.workspace.service.workspace.model.AzureCloudContext;
+import com.azure.resourcemanager.compute.ComputeManager;
+import com.azure.resourcemanager.relay.RelayManager;
+import com.azure.resourcemanager.storage.StorageManager;
+import com.fasterxml.jackson.core.type.TypeReference;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Retrieves the Azure resource's cloud region according to the resource type. */
+public class RetrieveAzureResourcesRegionStep implements Step {
+  private final Logger logger = LoggerFactory.getLogger(RetrieveAzureResourcesRegionStep.class);
+
+  private final AzureConfiguration azureConfig;
+  private final CrlService crlService;
+  private final ResourceDao resourceDao;
+  private final LandingZoneApiDispatch landingZoneApiDispatch;
+  private final AuthenticatedUserRequest userRequest;
+
+  public RetrieveAzureResourcesRegionStep(
+      AzureConfiguration azureConfig,
+      CrlService crlService,
+      ResourceDao resourceDao,
+      LandingZoneApiDispatch landingZoneApiDispatch,
+      AuthenticatedUserRequest userRequest) {
+    this.azureConfig = azureConfig;
+    this.crlService = crlService;
+    this.resourceDao = resourceDao;
+    this.landingZoneApiDispatch = landingZoneApiDispatch;
+    this.userRequest = userRequest;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    validateRequiredEntries(context.getWorkingMap(), CONTROLLED_RESOURCES_WITHOUT_REGION);
+    List<ControlledResource> controlledResources =
+        context.getWorkingMap().get(CONTROLLED_RESOURCES_WITHOUT_REGION, new TypeReference<>() {});
+    Map<UUID, String> workspaceIdToAzureCloudContextMap =
+        context
+            .getWorkingMap()
+            .get(
+                ControlledResourceKeys.WORKSPACE_ID_TO_AZURE_CLOUD_CONTEXT_MAP,
+                new TypeReference<>() {});
+    Map<UUID, String> resourceIdToRegionMap = new HashMap<>();
+    Map<UUID, String> resourceIdToWorkspaceIdMap = new HashMap<>();
+    for (var resource : controlledResources) {
+      WsmResourceType resourceType = resource.getResourceType();
+      logger.info(
+          "Getting cloud region for resource {} in workspace {}",
+          resource.getResourceId(),
+          resource.getWorkspaceId());
+      AzureCloudContext azureCloudContext =
```

Comment:
Is it possible there could be a disconnect between the `workspaceIdToAzureCloudContextMap` and the resources in `controlledResources` that would result in an NPE here?

Reply:
i don't think so?
cuz we iterate all the workspaces of the controlled resources and fetch the azure context (and throw if not found) in the previous step. 

however I can add a check here for best practice.